include/deal.II/matrix_free/evaluation_kernels.h

```diff
@@ -918,6 +995,587 @@ namespace internal
+                                                        hessians_quad,
+                                                        n_q_points);
+
+    if (dim == 3)
+      {
+        // advance to the next component in 1D array
+        values_dofs += dofs_per_comp;
+        values_quad += n_q_points;
+        gradients_quad += 3 * n_q_points;
+        hessians_quad += 6 * n_q_points;
+
+        // ******************************************
+        // *********** Third component **************
+        evaluate_tensor_product_per_component<2,
+                                              Eval_tangent,
+                                              Eval_tangent,
+                                              Eval_normal>(eval_tangent,
+                                                           eval_tangent,
+                                                           eval_normal,
+                                                           evaluation_flag,
+                                                           values_dofs,
+                                                           temp1,
+                                                           temp2,
+                                                           values_quad,
+                                                           gradients_quad,
+                                                           hessians_quad,
+                                                           n_q_points);
+      }
+  }
+
+
+  template <int dim, int fe_degree, int n_q_points_1d, typename Number>
+  template <int normal_dir, typename T0, typename T1, typename T2>
+  inline void
+  FEEvaluationImpl<MatrixFreeFunctions::tensor_raviart_thomas,
+                   dim,
+                   fe_degree,
+                   n_q_points_1d,
+                   Number>::
+    evaluate_tensor_product_per_component(
+      const T0                               eval0,
+      const T1                               eval1,
+      const T2                               eval2,
+      const EvaluationFlags::EvaluationFlags evaluation_flag,
+      const Number *                         values_dofs,
+      Number *                               temp1,
+      Number *                               temp2,
+      Number *                               values_quad,
+      Number *                               gradients_quad,
+      Number *                               hessians_quad,
+      const std::size_t                      n_q_points)
+  {
+    switch (dim)
+      {
+        case 2:
+          if (evaluation_flag & EvaluationFlags::gradients)
+            {
+              eval0.template gradients<0, true, false, normal_dir>(values_dofs,
+                                                                   temp1);
+              eval1.template values<1, true, false, normal_dir>(temp1,
+                                                                gradients_quad);
```

Comment:
That's really nice code, one can hardly see that there is a different polynomial space in the two directions. Out of curiosity, is the only difference the `normal_dir` argument to the other function? In that case, we might want to add the template argument there and merely use that function (of course, that would invalidate my other point above with pulling in `FEEvaluationData`, as we would then move those out).

Reply:
Yes it should be the exact same function, just without `normal_dir`. At first I actually added the template argument to `EvaluatorTensorProduct<evaluate_general,...>` but I realized something (that I can't remember now) that made me postpone this modification. But I think it is very doable and would remove a lot of lines.

Reply:
It would be great if you could try this, but if you don't manage I think we can go with the current state as well and try to clean things up later.

Reply:
Conceptually, I think `normal_dir` belong to `EvaluatorTensorProduct`. It is a parameter like `degree`, specifying the data  layout. In this anisotropic case, the `normal_dir` specified which degree is different; e.g., for `normal_dir=0`: `degree+1 x degree x degree`. I would create `EvaluatorTensorProduct` objects with the right template parameters locally.

Reply:
A general question: how does the data layout look like in the case of `Nedelec`? I am wondering if `normal_dir` is the right template parameter of the new `EvaluatorTensorProduct` class and we shouldn't explicitly write the degrees dim time?

Reply:
For Nedelec, things are mostly similar at the place of the tensor product evaluation we are looking at, with one minor difference: There "normal" direction has `degree`, whereas the tangential directions have `degree+1`. So the main difference is just how we interpret the degrees by some template arguments, whereas the actual code path is equivalent otherwise.

Reply:
I think we should unify the code paths for tensor product, RT, and Nedelec if possible (in the future). At the moment these evaluation calls look like copy-and-paste. But maybe the Piola transforms (as indicated by the todos below) will add here some additional complexity.

Reply:
The piola transform should be independent from the local evaluation of polynomials, as that is only a mapping effect (except for possibly copying in case we test by the gradient of test functions, similarly to the Hessian previously).